client/profile/user-context-menu.tsx

```diff
@@ -1,34 +1,79 @@
-import React, { useCallback } from 'react'
+import React, { useCallback, useEffect, useRef, useState } from 'react'
 import styled from 'styled-components'
+import { useRoute } from 'wouter'
+import { ChannelModerationAction } from '../../common/chat'
+import { MULTI_CHANNEL } from '../../common/flags'
 import { SbUserId } from '../../common/users/sb-user'
+import { useSelfUser } from '../auth/state-hooks'
+import { getChatUserProfile, moderateUser } from '../chat/action-creators'
+import { openDialog } from '../dialogs/action-creators'
+import { DialogType } from '../dialogs/dialog-type'
+import { Divider } from '../material/menu/divider'
 import { MenuItem } from '../material/menu/item'
 import { Menu } from '../material/menu/menu'
 import { PopoverProps } from '../material/popover'
 import { inviteToParty, kickPlayer, removePartyInvite } from '../parties/action-creators'
 import { useAppDispatch, useAppSelector } from '../redux-hooks'
+import { openSnackbar } from '../snackbars/action-creators'
 import { colorTextFaint } from '../styles/colors'
+import { subtitle1 } from '../styles/typography'
 import { navigateToWhisper } from '../whispers/action-creators'
 import { navigateToUserProfile } from './action-creators'
 
 const LoadingItem = styled(MenuItem)`
   color: ${colorTextFaint};
 `
 
+const LoadingError = styled.div`
+  ${subtitle1};
+  padding: 16px;
+`
+
 export interface ConnectedUserContextMenuProps {
   userId: SbUserId
   popoverProps: Omit<PopoverProps, 'children'>
 }
 
 export function ConnectedUserContextMenu({ userId, popoverProps }: ConnectedUserContextMenuProps) {
   const dispatch = useAppDispatch()
-  const selfUser = useAppSelector(s => s.auth.user)
+  const selfUser = useSelfUser()
+  const cancelLoadRef = useRef(new AbortController())
+  const [isLoadingChatUserProfile, setIsLoadingChatUserProfile] = useState<boolean>(false)
+  const [loadingChatUserProfileError, setLoadingChatUserProfileError] = useState<Error>()
+  const selfPermissions = useAppSelector(s => s.auth.permissions)
   const user = useAppSelector(s => s.users.byId.get(userId))
+  const [, routeParams] = useRoute('/chat/:channelName')
```

Comment:
if we're using this in a bunch of places, it should be declared as a hook somewhere. This is gonna be hard to track down if we e.g. change the chat paths at some point in the future

Reply:
This also, imo, doesn't really make much sense here. This component is not only used inside chat channels. The intended path for special menu items that depend on context was more like the handling in my shift-click change. I didn't yet build the similar stuff for menu items, but you should do that here instead of checking the route at all.

Reply:
I'm not sure I quite understand how to do that. I took a look at the commit where you added the shift-click change, but I only ended up even more confused :p

Reply:
You need a way of passing in extra menu items, or modifying the list of menu items, or something along those lines. You can do it as props directly, you can use Context for it (as in the shift-click stuff), etc. There's a lot of options. What you *can't* do is what you've done here: add super-specific logic to a very general component